doc/adrs/027_make_the_explorer_independent_from_modeling_concepts.adoc

```diff
@@ -0,0 +1,76 @@
+= ADR-027 - Make the explorer independent from modeling concepts
+
+== Context
+
+The explorer is hardcoded with some of our own modeling concepts and it can only really be used with our own dedicated concepts.
+As an example, the following concepts are hardcoded in the `TreeItem`:
+
+- create, rename, delete representation
+- create, rename, delete object
+- create, rename, delete, download document
+- TreeItemDiagramContextMenu (used for all representations including the Form one...)
+
+This prevent the explorer from being reused for other use cases.
+
+== Decision
+
+We will make the explorer more generic and it will act a bit more like other representations with generic tools which can be mapped to concrete concepts known only by the backend.
+We will start by providing generic actions with handlers that will need to be provided just like for other representations.
+
+=== Support of generic actions like renaming and deletion
+
+In order to support the deletion and renaming of tree item, new properties `TreeItem#deletable` and `TreeItem#renamable` will be added.
+When available, the frontend will provide support for the deletion and renaming of said tree item.
+It will also call some generic mutation such as `Mutation#deleteTreeItem` and `Mutation#renameTreeItem` to perform the action.
+Sirius Components does not define what should appear in the explorer and the project integrating Sirius Components will have to handle the deletion and renaming of the concept behind the tree item.
+
+As such, Sirius Web will have to include both a `DeleteTreeItemEventHandler to handle the deletion of the requested tree item and and `RenameTreeItemEventHandler` for the renaming.
+It should then delegate to the renaming or deletion of a document, representation or even an object to the relevant service but that's an internal concern of Sirius Web.
+
+=== Support of specific actions such as child creation and download
+
+The creation of a new child may require specific information, for example:
+
+- The URI of the domain and the name of the entity to create a new child object
+- The name of the representation along with its type
+- The stereotype of its document and its name
+
+Those various use cases may require interaction with a server and some custom user interface.
+The same concerns are relevant to support downloading documents today (custom user interface + custom endpoint).
+
+In order to support those use cases, we will need to have the ability to contribute custom entries in the context menus of the TreeItem.
+In a similar fashion as the ability to contribute additional representations, the Sirius Components frontend API will provide the ability to add entries in the context menu.
+Several examples will be provided by Sirius Web:
+
+- A direct download link to download a document
+- An entry to create new objects thanks to the `NewObjectModal`
```

Comment:
You wrote right above that "The creation of a new child may require specific information". How will the `NewObjectModal` ask for this specific information?

Reply:
All I am saying is that creating child right now requires a dedicated user interface with some dedicated GraphQL endpoints etc. Thus, we can't and won't provide a generic way to create child but instead let anyone provide their own user interface to provide, for example, the current NewObjectModal, the NewRepresentationModal, etc. Right now, the NewObjectModal ask for the child creation descriptions for example.